maven-core/src/main/java/org/apache/maven/execution/LockingEventSpy.java

```diff
@@ -0,0 +1,96 @@
+package org.apache.maven.execution;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.maven.eventspy.AbstractEventSpy;
+import org.apache.maven.eventspy.EventSpy;
+import org.apache.maven.project.MavenProject;
+import org.eclipse.aether.SessionData;
+import org.eclipse.sisu.Typed;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * EventSpy implementation that provides a per-project locking mechanism
+ * to make sure a given project can not be build twice concurrently.
+ * This case can happen when running parallel builds with forked lifecycles.
+ */
+@Singleton
+@Named
+@Typed( EventSpy.class )
+public class LockingEventSpy extends AbstractEventSpy
```

Comment:
I think it's quite an elegant solution, but I'm not sure if using an `EventSpy` for this purpose is how it should be done in `maven-core`. I don't think there are other implementations of it in core.

Reply:
Another possibility would be to change the `MavenCli` to support multiple `ExecutionListener`.  The problem (minor) is that there's no discovery mechanism, so they are hard coded, which I wanted to avoid...  Another possibility would be to define a component to hold the locks and inject it into the `LifeycleModuleBuilder` and `MojoExecutor` where those events are fired initially.  I don't really mind, but the current solution was the less intrusive.

Reply:
It's definitely less intrusive, but I'm not sure whether something that controls the build as much as this (temporarily pausing and resuming project builds) should be done in an `ExecutionListener` or `EventSpy`. My gut feeling says those classes should only extend the build with simple side-effects, like logging a project's phase. I would prefer your last suggestion. Then again, maybe this is the right direction. Perhaps some others can give their opinion on it?